### PR TITLE
Debugger pending breakpoints

### PR DESCRIPTION
**Pull request by kisbg**
Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

File: src/debugger/Debugger.h [hunk 1]
```diff
@@ -265,6 +277,8 @@ class Debugger : public gc {
 
     uint8_t m_delay;
     bool m_computeLocation;
+    bool m_pendingWait;
+    bool m_waitForResume;
```
Comment (zherczeg) [inline src/debugger/Debugger.h:281]: Perhaps `:1` could be added to these bools.

File: tools/debugger/debugger_core.py [hunk 2]
```diff
@@ -70,6 +70,8 @@
 ESCARGOT_MESSAGE_EXCEPTION = 41
 ESCARGOT_MESSAGE_EXCEPTION_BACKTRACE = 42
 ESCARGOT_DEBUGGER_WAIT_FOR_SOURCE = 43
+ESCARGOT_DEBUGGER_WAITING_AFTER_PENDING = 44
+
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:74]: Remove this newline.

File: tools/debugger/debugger_core.py [hunk 3]
```diff
@@ -844,6 +906,19 @@ def _send_breakpoint(self, breakpoint):
...
+        if enable:
+            message = struct.pack(self.byte_order + "BBB",
+                                  1 + 1,
+                                  ESCARGOT_DEBUGGER_PENDING_CONFIG_ENABLE,
+                                  enable)
+        else:
+            message = struct.pack(self.byte_order + "BBB",
+                                  1 + 1,
+                                  ESCARGOT_DEBUGGER_PENDING_CONFIG_DISABLE,
+                                  enable)
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:919]: I don't get this. We have `ESCARGOT_DEBUGGER_PENDING_CONFIG_ENABLE` / `ESCARGOT_DEBUGGER_PENDING_CONFIG_DISABLE` but we also send `enable` as well? Perhaps we could have only a single `ESCARGOT_DEBUGGER_PENDING_CONFIG` message then.

File: tools/debugger/debugger_core.py [hunk 4]
```diff
@@ -976,3 +1077,24 @@ def _receive_string(self, message_type, data):
...
+    def delete_pending(self):
+        if self.pending_breakpoint_list:
+            self.pending_breakpoint_list.clear()
+            self._send_pending_config(0)
+
+    def breakpoint_pending_exists(self, breakpoint):
+        for existing_bp in self.pending_breakpoint_list.values():
+            if (breakpoint.line and existing_bp.source_name == breakpoint.source_name and \
+                  existing_bp.line == breakpoint.line) \
+               or (not breakpoint.line and existing_bp.function == breakpoint.function):
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:1097]: This looks misaligned. Would be good if we could make this condition nicer.

File: tools/debugger/tests/do_pending_breakpoints_lines.expected [hunk 5]
```diff
@@ -0,0 +1,22 @@
+Connecting to: localhost:6501
+Connection created!!!
+Stopped at tools/debugger/tests/do_pending_breakpoints_lines.js:20
+(escargot-debugger) b :1
```
Comment (zherczeg) [inline tools/debugger/tests/do_pending_breakpoints_lines.expected:4]: I don't see a pending breakpoint with a file name. Am I missing it?

**Comment by kisbg** [comment]
@zherczeg  Thank you for the review!

File: src/debugger/Debugger.cpp [hunk 6]
```diff
@@ -686,16 +686,44 @@ bool Debugger::processIncomingMessages(ExecutionState* state, ByteCodeBlock* byt
             getScopeVariables(state, index);
             return true;
         }
+        case ESCARGOT_DEBUGGER_PENDING_CONFIG: {
+            if (length != 2) {
+                break;
+            }
+            uint8_t index;
+            memcpy(&index, buffer + 1, sizeof(uint8_t));
```
Comment (zherczeg) [inline src/debugger/Debugger.cpp:694]: there is no need any memcpy here, just `m_pendingWait = buffer[1]`.

File: tools/debugger/debugger_core.py [hunk 7]
```diff
@@ -844,6 +904,13 @@ def _send_breakpoint(self, breakpoint):
                               breakpoint.offset)
         self.channel.send_message(self.byte_order, message)
 
+    def _send_pending_config(self, enable):
+        message = struct.pack(self.byte_order + "BBb",
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:908]: `BBb`? I think it should be ``BBB`

**Pull request merged**